apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/AplDbVersion.java

```diff
@@ -688,38 +688,52 @@ protected int update(int nextUpdate) {
             case 257:
+                apply("CREATE TABLE IF NOT EXISTS transaction_shard_index (transaction_id BIGINT NOT NULL, partial_transaction_hash VARBINARY NOT NULL, block_id BIGINT NOT NULL)");
+            case 258:
                 apply("ALTER TABLE transaction_shard_index ADD CONSTRAINT IF NOT EXISTS fk_transaction_shard_index_block_id " +
                         "FOREIGN KEY (block_id) REFERENCES block_index(block_id) ON DELETE CASCADE");
-            case 258:
-                apply("CREATE UNIQUE INDEX IF NOT EXISTS transaction_index_shard_1_idx ON transaction_shard_index (transaction_id, block_id)");
             case 259:
+                apply("CREATE UNIQUE INDEX IF NOT EXISTS transaction_index_shard_1_idx ON transaction_shard_index (transaction_id, block_id)");
+            case 260:
+                apply("CREATE TABLE IF NOT EXISTS referenced_shard_transaction (db_id BIGINT auto_increment NOT NULL, transaction_id BIGINT NOT NULL, " +
+                        "referenced_transaction_id BIGINT NOT NULL)");
+            case 261:
+                apply("ALTER TABLE referenced_shard_transaction ADD CONSTRAINT IF NOT EXISTS pk_referenced_shard_transaction_db_id PRIMARY KEY(db_id)");
+            case 262:
+                apply("ALTER TABLE referenced_shard_transaction ADD CONSTRAINT IF NOT EXISTS " +
+                        "fk_referenced_shard_transaction_transaction_id_transaction_shard_index_transaction_id " +
+                        "FOREIGN KEY (transaction_id) REFERENCES transaction_shard_index (transaction_id) ON DELETE CASCADE");
+            case 263:
                 apply("CREATE TABLE IF NOT EXISTS shard_recovery (shard_recovery_id BIGINT AUTO_INCREMENT NOT NULL, " +
                         "state VARCHAR NOT NULL, object_name VARCHAR NULL, column_name VARCHAR NULL, " +
                         "last_column_value BIGINT, processed_object VARCHAR, updated TIMESTAMP(9) NOT NULL)");
-            case 260:
+            case 264:
                 apply("ALTER TABLE shard_recovery ADD CONSTRAINT IF NOT EXISTS pk_shard_recovery_state PRIMARY KEY(shard_recovery_id)");
-            case 261:
+            case 265:
                 apply("ALTER TABLE shard_recovery ADD CONSTRAINT IF NOT EXISTS shard_recovery_id_state_object_idx unique (shard_recovery_id, state)");
-            case 262:
+            case 266:
+                apply("ALTER TABLE shard_recovery ADD CONSTRAINT IF NOT EXISTS shard_recovery_id_state_object_idx unique (shard_recovery_id, state)");
+            case 267:
                 apply("ALTER TABLE genesis_public_key DROP CONSTRAINT IF EXISTS CONSTRAINT_C11");
-            case 263:
-                apply("ALTER TABLE referenced_transaction ADD COLUMN height INT NOT NULL");
-            case 264:
+            case 268:
+                apply("ALTER TABLE IF EXISTS referenced_transaction ADD COLUMN IF NOT EXISTS height INT NOT NULL");
```

Comment:
referenced_transaction table should exist here

Reply:
That variant is more safe, if it's needed.